### ISSUE TITLE
Introduce grpc health check in etcd client

### ISSUE THREAD

**Issue by chaochn47** [etcd-io/etcd#16276]
### Background

> gRPC Health checks are used to probe whether the server is able to handle RPCs. A server may choose to reply “unhealthy” because it is not ready to take requests, it is shutting down or some other reason. The client can act accordingly if the response is not received within some time window or the response says unhealthy in it.

ref. 
1. https://github.com/grpc/grpc/blob/master/doc/health-checking.md
2. https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md

https://github.com/etcd-io/etcd/pull/8121 added basic grpc health service only on server side since etcd v3.3. 
https://github.com/etcd-io/etcd/blob/622017468725647028e75aea2c971966ab70b3b5/server/etcdserver/api/v3rpc/grpc.go#L81-L86

### Problem

In a multi etcd server endpoints scenario, etcd client only fails over to the other endpoint when existing connection/channel is in not in `Ready` state. However, etcd client does not know about if etcd server can handle RPCs. 

For example
1. https://github.com/etcd-io/etcd/issues/9222
2. https://github.com/etcd-io/etcd/issues/14338

It needs a comprehensive design and testing. 

Placeholder google doc [etcd client grpc health check](https://docs.google.com/document/d/1x7guECGQI1wdAh2fBzPBRTva-sqGGLR8Bp0lrrmmDk0/edit#heading=h.euv7j5s250nm) copied from the [KEP template](https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template). 

### Why is this needed?

Improve etcd availability by failing over to other healthy etcd endpoints

**Comment by fuweid** [comment]
Hi @chaochn47 , based on your PR #16278, it seems like that the PR marks not_serving status because of defrag. It should be applied to any maintenance APIs, is it correct? My only concern is that based on that https://github.com/grpc/grpc-go/blob/d7f45cdf9ae720256d328fcdb6356ae58122a7f6/health/client.go#L104, it might impact the on-going requests(still in trasfering data) and force to recreate connection. (I haven't verified it yet,  update it later)

**Comment by chaochn47** [comment]
Thanks @fuweid for the review!

> based on your PR https://github.com/etcd-io/etcd/pull/16278, it seems like that the PR marks not_serving status because of defrag. It should be applied to any maintenance APIs, is it correct?

Online Defrag is one of the use cases that server reports "unhealthy" to client to fail over the requests. I do not plan to support any other maintenance APIs now. https://github.com/etcd-io/etcd/pull/16278 is just a POC and health server should not exposed to maintenance server, while to defrag API only ideally. 

> My only concern is that based on that https://github.com/grpc/grpc-go/blob/d7f45cdf9ae720256d328fcdb6356ae58122a7f6/health/client.go#L104, it might impact the on-going requests(still in trasfering data) and force to recreate connection. 

I won't be too concerned about this failure mode since it is not called out in https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md and https://github.com/grpc/proposal/pull/97. But I will validate with adding some testing cases in https://github.com/grpc/grpc-go/blob/master/test/healthcheck_test.go.

**Comment by serathius** [comment]
> However, gRPC client side health check replies on etcd (gRPC) server pushes its health status to client for requests fail-over via a server side streaming API. It is designed only for round-robin policy and will improve etcd availability.

Pull push, just changes control over validation period. Server or client. 

What's important here is definition of etcd health. I expect both types of health checking will end up with pretty similar definition of health.

**Comment by chaochn47** [comment]
Thanks. I think we are on the same page now. 

Do you still think it's a duplicate of https://github.com/etcd-io/etcd/issues/16007? Will it cover how to consume the health status change? 

I would like the etcd client to fail over to other healthy endpoints considering etcd health, not just connection is up or down.

**Comment by wenjiaswe** [comment]
> gRPC Health checks are used to probe whether the server is able to handle RPCs.

That sounds like one case of not ready for /readyz, is my understanding correct?

**Comment by chaochn47** [comment]
Yeah I think so but I want this thread focus on etcd client fail over by introducing gRPC health checks. 

I suggest we move the conversation about etcd failure modes classification into which probes to https://github.com/etcd-io/etcd/issues/16007

**Comment by serathius** [comment]
This issue can be left opened, as a followup to introduce grpc client health checks when we figure out proper health definitions.

**Comment by ahrtr** [comment]
It's a valid feature to me, and I agree we should design it together with the [Livez/Readyz](https://github.com/etcd-io/etcd/issues/16007) feature, because the key is how to define health (e.g. get stuck on write for a long time > a threshold).

Also two detailed comments:
- We plan to get rid of defragmentation in 3.7 in https://github.com/etcd-io/bbolt/issues/422.  [But I do not get time to do a PoC so far :( ]
- The PR https://github.com/etcd-io/etcd/pull/16278 is still depending on an experimental gRPC API `WithResolvers`, but we want to remove the dependency on the experimental gRPC API in https://github.com/etcd-io/etcd/issues/15145

**Comment by chaochn47** [comment]
Thanks @ahrtr for the review, 

Definitely we want any feature request to be future-proof and won't be thrown away. However, unless we have a foreseeable approach to resolve the above two items in the upcoming year, this issue can be left on the table.

**Comment by chaochn47** [comment]
> My only concern is that based on that https://github.com/grpc/grpc-go/blob/d7f45cdf9ae720256d328fcdb6356ae58122a7f6/health/client.go#L104, it might impact the on-going requests(still in trasfering data) and force to recreate connection. (I haven't verified it yet, update it later)

Hi @fuweid, I think the on-going RPCs can still proceed, please check the simulated test case https://github.com/chaochn47/grpc-go/pull/1. Thank you ~

**Comment by chaochn47** [comment]
> One caveat is if the watch stream is not interrupted, client that is built on top of the watch as a cache may be stale. In order to terminate the watch stream, I guess the etcd server needs to be stopped.

FWIW, when the disk I/O is blocked/stalled, partitioned node stays with state `s.Leader() != types.ID(raft.None)`. Watch is not cancelled because of [monitorLeader](https://github.com/etcd-io/etcd/blob/9e1abbab6e4ba2886238f49b0d48fc19a546c7cf/server/etcdserver/api/v3rpc/interceptor.go#L313) depends on `s.Leader() == types.ID(raft.None)`. It is explained in https://github.com/etcd-io/etcd/issues/13527 and reproduce steps provided. 

kube-apiserver watch cache built on top of etcd watch will not receive update and could potentially receive stale data. Instead, the watch should be cancelled by server and requires client to re-create watch on a different etcd node that is connected to quorum.

So after etcd health status is properly defined, it should be consumed by `monitorLeader` as well.

**Comment by lavacat** [comment]
I think this is very useful specifically for defrag. In high db size, high traffic clusters this is a real problem. When `maxGapBetweenApplyAndCommitIndex` is reached and defrag is still blocking applies, etcd will fail requests.
